app/services/slack_service.rb

```diff
@@ -0,0 +1,200 @@
+
+    if message == nil || message == ""
+      raise InvalidCommand.new 'Did you include a message surrounded by \'?'
+    end
+
+    amount = get_amount_from_command(command_text)
+
+    begin
+      PostCreator.create_post(message, amount, sender, receivers, team)
+    rescue PostCreator::PostCreateError => e
+      raise InvalidCommand.new e
+    end
+  end
+
+  def self.send_post_announcement(post)
+    client = Slack::Web::Client.new(token: post.team.slack_bot_access_token)
+
+    sender_string = post.sender.slack_id == nil ? "#{post.sender.name}" : "<@#{post.sender.slack_id}>"
+
+    message = "#{sender_string} just gave #{post.amount} kudos to #{get_post_receivers(post)} for #{post.message}"
+
+    client.chat_postMessage(channel: post.team.channel_id, text: message)
+  end
+
+  def self.connect_account(command_text, user_id)
+    slack_register_token = command_text
+
+    raise InvalidCommand.new 'please provide a register token' unless slack_register_token != ''
+
+    raise InvalidCommand.new 'This slack account is already linked to kudo-o-matic' unless User.where(:slack_id => user_id).count == 0
+
+    user = User.where(:slack_registration_token => slack_register_token).take
+
+    raise InvalidCommand.new 'Invalid registration token' unless user != nil
+
+    raise InvalidCommand.new 'This kudo-o-matic account is already linked to Slack.' unless user.slack_id == nil
+
+    user.slack_id = user_id
+    user.slack_registration_token = nil
+    raise InvalidCommand.new "That didn't quite work, #{user.errors.full_messages.join(', ')}" unless user.save
+  end
+
+  def self.add_to_workspace(code, team_id)
+    client = Slack::Web::Client.new
+    if code == nil
+      raise InvalidRequest.new 'Auth token is missing'
+    end
+
+    if team_id == nil
+      raise InvalidRequest.new 'Team id is missing'
+    end
+    team = Team.find(team_id)
+
+    rc = client.oauth_v2_access(
+        code: code,
+        client_id: ENV['SLACK_CLIENT_ID'],
+        client_secret: ENV['SLACK_CLIENT_SECRET']
+    )
+
+    team.channel_id = rc['incoming_webhook']['channel_id']
+    team.slack_bot_access_token = rc["access_token"]
+    team.slack_team_id = rc["team"]["id"]
+
+    raise InvalidRequest.new "That didn't quite work, #{team.errors.full_messages.join(', ')}" unless team.save
+
+    send_welcome_message(team)
+  end
+
+  def self.list_guidelines(slack_team_id)
+    team = Team.find_by_slack_team_id(slack_team_id)
+
+    raise InvalidCommand.new 'No team with that Slack ID' unless team != nil
+
+    guidelines = Guideline.where(:team => team).order(:kudos)
+
+    if guidelines.count == 0
+      return create_markdown_block("No guidelines")
+    else
+      return create_markdown_block(guidelines_to_list(guidelines))
+    end
+  end
+
+  def self.get_oauth_url(team_id)
+    URI::HTTP.build(
+        :host => Settings.slack_auth_endpoint,
+        :path => '/oauth/v2/authorize',
+        :query => {
+            :state => team_id,
```

Comment:
I would prefer you not abuse the OAuth state parameter for this. I think you can specify a `redirect_uri` for Slack to redirect the user after authentication, you could stick it on there?